netutils/config/parser.py

```diff
@@ -1419,3 +1421,99 @@ def _build_banner(self, config_line: str) -> t.Optional[str]:
+        """Demarcate End of Banner char(s)."""
+        raise NotImplementedError("Palo Alto Networks platform doesn't have a end of banner character.")
+
+    def is_banner_end(self, line: str) -> bool:
+        """Determine if end of banner."""
+        if line.endswith('"') or line.startswith('";') or line.startswith("set"):
+            return True
+        return False
+
+    def _build_banner(self, config_line: str) -> t.Optional[str]:
+        """Handle banner config lines.
+
+        Args:
+            config_line: The start of the banner (system note) config.
+
+        Returns:
+            The next configuration line in the configuration text or None when banner end is the end of the config text.
+
+        Raises:
+            ValueError: When the parser is unable to identify the End of the Banner.
+        """
+        banner_config = [config_line]
+        for line in self.generator_config:
+            if not self.is_banner_end(line):
+                banner_config.append(line)
+            else:
+                banner_config.append(line)
+                line = "\n".join(banner_config)
+                self._update_config_lines(line)
+                try:
+                    return next(self.generator_config)
+                except StopIteration:
+                    return None
+        raise ValueError("Unable to parse banner end.")
+
+    def build_config_relationship(self) -> t.List[ConfigLine]:
+        r"""Parse text of config lines and find their parents.
+
+        Examples:
+            >>> config = (
+            ...     "set devices localhost.localdomain deviceconfig system hostname firewall1\n"
+            ...     "set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server 10.0.0.1\n"
+            ...     "set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server-2 10.0.0.2\n"
+            ...     "set devices localhost.localdomain deviceconfig setting config rematch yes\n"
+            ... )
+            >>> config_tree = PaloAltoNetworksConfigParser(config)
+            >>> config_tree.build_config_relationship() == \
+            ... [
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig system hostname firewall1", parents=()),
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server 10.0.0.1", parents=()),
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig system panorama local-panorama panorama-server-2 10.0.0.2", parents=()),
+            ...     ConfigLine(config_line="set devices localhost.localdomain deviceconfig setting config rematch yes", parents=()),
+            ... ]
+            True
+        """
+        # if config is in palo/json format, convert to set
+        if self.config_lines_only.startswith("config {") and self.config_lines_only is not None:
```

Comment:
My gut says this should be a bit more resilient, such as if the second line has "config {" or something similar.

Reply:
Modifying this to check for any lines in the config that end in { would trigger conversion

Reply:
Modifying this to check for any lines in the config that end in { would trigger conversion